### PR TITLE
Remove the futures from prefetch

### PR DESCRIPTION
**Pull request by mykhailo-kuchma**
Add a PrefetchJob class to help track the prefetch progress
Remove the usage of std::future and additional task that wait for
them.

Relates-To: OLPEDGE-2169

Signed-off-by: Mykhailo Kuchma <ext-mykhailo.kuchma@here.com>

### CODE REVIEW

File: olp-cpp-sdk-dataservice-read/src/repositories/PrefetchJob.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  prefetch_result_.push_back(std::move(result));
+
+  if (!--task_count_) {
+    PrefetchTilesResponseCallback user_callback = std::move(user_callback_);
+    if (canceled_) {
+      user_callback({{client::ErrorCode::Cancelled, "Cancelled"}});
+    } else {
+      user_callback(std::move(prefetch_result_));
+    }
+  }
```
Comment (kzvieriev) [inline olp-cpp-sdk-dataservice-read/src/repositories/PrefetchJob.cpp:75]: We can get into trouble, if first N tasks completed before N+1 task added to the `PrefetchJob`.

File: olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp [hunk 2]
```diff
@@ -271,117 +273,52 @@ client::CancellationToken VersionedLayerClientImpl::PrefetchTiles(
...
+                  settings.task_scheduler, pending_requests,
+                  [=](CancellationContext inner_context) {
+                    repository::DataCacheRepository data_cache_repository(
+                        catalog, shared_settings->cache);
+                    if (data_cache_repository.IsCached(layer_id, handle)) {
+                      // Cached, return an empty success
+                      return DataResponse(nullptr);
+                    } else {
+                      DataRequest request;
+                      request.WithDataHandle(handle).WithBillingTag(biling_tag);
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/VersionedLayerClientImpl.cpp:302]:
```suggestion
                      auto request = DataRequest().WithDataHandle(handle).WithBillingTag(biling_tag);
```

File: olp-cpp-sdk-dataservice-read/src/repositories/PrefetchJob.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    : user_callback_(std::move(user_callback)),
+      task_count_{task_count},
+      canceled_{false} {
+  tasks_contexts_.reserve(task_count_);
+  prefetch_result_.reserve(task_count_);
+}
+
+PrefetchJob::~PrefetchJob() {
+  assert(task_count_ == 0);
+  assert(user_callback_ == nullptr);
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PrefetchJob.cpp:47]: No asserts!

File: olp-cpp-sdk-dataservice-read/src/repositories/PrefetchJob.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+#include <olp/dataservice/read/PrefetchTileResult.h>
+
+namespace {
+constexpr auto kLogTag = "PrefetchJob";
+}
+
+namespace olp {
+namespace dataservice {
+namespace read {
+namespace repository {
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PrefetchJob.cpp:34]: Doesn't seem like it fits in `repository` to me.

File: olp-cpp-sdk-dataservice-read/src/repositories/PrefetchJob.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+  void CompleteTask(geo::TileKey tile);
+
+  void CompleteTask(geo::TileKey tile, const client::ApiError& error);
+
+  void CancelOperation();
+
+ protected:
+  void CompleteTask(std::shared_ptr<PrefetchTileResult> result);
+
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/repositories/PrefetchJob.h:52]: private:

File: olp-cpp-sdk-dataservice-read/src/PrefetchJob.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#include "PrefetchJob.h"
+
+#include <assert.h>
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/PrefetchJob.cpp:22]: Remove

**Pull request merged**